### PR TITLE
[debug dump util] COPP Module Added

### PR DESCRIPTION
**Pull request by vivekreddynv**
Signed-off-by: Vivek Reddy Karri <vkarri@nvidia.com>

#### What I did
HLD for Dump Utility: [HLD](https://github.com/Azure/SONiC/pull/789).
* Added the COPP Module to the Debug Dump Utility
* Added the Corresponding UT's

#### How to verify it

```
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_all_args PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_custom_trap_with_default_grp PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_custom_trap_with_missing_group PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_default_grp_and_custom_trap_with_diff PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_default_trap_and_default_grp PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_default_trap_and_default_grp_with_diff PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_invalid_trap_id PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_missing_appl PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_missing_asic_dump PASSED [ 85%]
tests/dump_tests/module_tests/copp_test.py::TestCoppModule::test_missing_state PASSED [ 86%]
```

#### New command output (if the output of a command-line utility has changed)

```
from dump.plugins.copp import Copp 
m_copp = Copp()
params = {Copp.ARG_NAME : "sample_packet", "sample_packet"  : "" }
dict = m_copp.execute(params)
print(dict)

{
    "CONFIG_DB": {
        "keys": [
            "COPP_GROUP|queue2_group1"
        ],
        "tables_not_found": []
    },
    "APPL_DB": {
        "keys": [
            "COPP_TABLE:queue2_group1"
        ],
        "tables_not_found": []
    },
    "ASIC_DB": {
        "keys": [
            "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP:oid:0x220000000004de",
            "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP_GROUP:oid:0x110000000004db",
            "ASIC_STATE:SAI_OBJECT_TYPE_POLICER:oid:0x120000000004dc",
            "ASIC_STATE:SAI_OBJECT_TYPE_QUEUE:oid:0x150000000002a1",
            "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TABLE_ENTRY:oid:0x230000000004d8",
            "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF:oid:0xd0000000004d6"
        ],
        "tables_not_found": []
    },
    "STATE_DB": {
        "keys": [
            "COPP_TRAP_TABLE|sflow",
            "COPP_GROUP_TABLE|queue2_group1"
        ],
        "tables_not_found": []
    },
    "CONFIG_FILE": {
        "keys": [
            "COPP_TRAP|sflow",
            "COPP_GROUP|queue2_group1"
        ],
        "tables_not_found": []
    }
}
```

```
admin@sonic:~$ dump state copp bgp
{
    "bgp": {
        "CONFIG_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "APPL_DB": {
            "keys": [
                {
                    "COPP_TABLE:queue4_group1": {
                        "queue": "4",
                        "trap_action": "trap",
                        "trap_ids": "bgp,bgpv6,lacp",
                        "trap_priority": "4"
                    }
                }
            ],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP:oid:0x22000000000c56": {
                        "SAI_HOSTIF_TRAP_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_TRAP",
                        "SAI_HOSTIF_TRAP_ATTR_TRAP_GROUP": "oid:0x11000000000c55",
                        "SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE": "SAI_HOSTIF_TRAP_TYPE_BGP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP_GROUP:oid:0x11000000000c55": {
                        "SAI_HOSTIF_TRAP_GROUP_ATTR_QUEUE": "4"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000626": {
                        "NULL": "NULL",
                        "SAI_QUEUE_ATTR_INDEX": "4",
                        "SAI_QUEUE_ATTR_TYPE": "SAI_QUEUE_TYPE_UNICAST"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x22000000000c56": "oid:0x400300000022",
                "oid:0x11000000000c55": "oid:0x200000011",
                "oid:0x15000000000626": "oid:0x12e0000040015"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "COPP_TRAP_TABLE|bgp": {
                        "state": "ok"
                    }
                },
                {
                    "COPP_GROUP_TABLE|queue4_group1": {
                        "state": "ok"
                    }
                }
            ],
            "tables_not_found": []
        },
        "CONFIG_FILE": {
            "keys": [
                {
                    "COPP_TRAP|bgp": {
                        "trap_ids": "bgp,bgpv6",
                        "trap_group": "queue4_group1"
                    }
                },
                {
                    "COPP_GROUP|queue4_group1": {
                        "trap_action": "trap",
                        "trap_priority": "4",
                        "queue": "4"
                    }
                }
            ],
            "tables_not_found": []
        }
    }
}


admin@sonic:~$ dump state copp arp_req --table --key-map
+-----------+-------------+-----------------------------------------------------------------------+
| trap_id   | DB_NAME     | DUMP                                                                  |
+===========+=============+=======================================================================+
| arp_req   | CONFIG_DB   | +------------------+                                                  |
|           |             | | Keys Collected   |                                                  |
|           |             | +==================+                                                  |
|           |             | +------------------+                                                  |
+-----------+-------------+-----------------------------------------------------------------------+
| arp_req   | APPL_DB     | +--------------------------+                                          |
|           |             | | Keys Collected           |                                          |
|           |             | +==========================+                                          |
|           |             | | COPP_TABLE:queue4_group2 |                                          |
|           |             | +--------------------------+                                          |
+-----------+-------------+-----------------------------------------------------------------------+
| arp_req   | ASIC_DB     | +-------------------------------------------------------------------+ |
|           |             | | Keys Collected                                                    | |
|           |             | +===================================================================+ |
|           |             | | ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP:oid:0x22000000000c5b       | |
|           |             | +-------------------------------------------------------------------+ |
|           |             | | ASIC_STATE:SAI_OBJECT_TYPE_HOSTIF_TRAP_GROUP:oid:0x11000000000c59 | |
|           |             | +-------------------------------------------------------------------+ |
|           |             | | ASIC_STATE:SAI_OBJECT_TYPE_POLICER:oid:0x12000000000c5a           | |
|           |             | +-------------------------------------------------------------------+ |
|           |             | | ASIC_STATE:SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000626             | |
|           |             | +-------------------------------------------------------------------+ |
|           |             | +----------------------+---------------------+                        |
|           |             | | vid                  | rid                 |                        |
|           |             | +======================+=====================+                        |
|           |             | | oid:0x22000000000c5b | oid:0x200000000022  |                        |
|           |             | +----------------------+---------------------+                        |
|           |             | | oid:0x11000000000c59 | oid:0x300000011     |                        |
|           |             | +----------------------+---------------------+                        |
|           |             | | oid:0x12000000000c5a | oid:0x200000012     |                        |
|           |             | +----------------------+---------------------+                        |
|           |             | | oid:0x15000000000626 | oid:0x12e0000040015 |                        |
|           |             | +----------------------+---------------------+                        |
+-----------+-------------+-----------------------------------------------------------------------+
| arp_req   | STATE_DB    | +--------------------------------+                                    |
|           |             | | Keys Collected                 |                                    |
|           |             | +================================+                                    |
|           |             | | COPP_TRAP_TABLE|arp            |                                    |
|           |             | +--------------------------------+                                    |
|           |             | | COPP_GROUP_TABLE|queue4_group2 |                                    |
|           |             | +--------------------------------+                                    |
+-----------+-------------+-----------------------------------------------------------------------+
| arp_req   | CONFIG_FILE | +--------------------------+                                          |
|           |             | | Keys Collected           |                                          |
|           |             | +==========================+                                          |
|           |             | | COPP_TRAP|arp            |                                          |
|           |             | +--------------------------+                                          |
|           |             | | COPP_GROUP|queue4_group2 |                                          |
|           |             | +--------------------------+                                          |
+-----------+-------------+-----------------------------------------------------------------------+

```

### CODE REVIEW

**Comment by prsunny** [comment]
@dgsudharsan , can you review/sign-off?

**Review by dgsudharsan — DISMISSED** [review]

**Comment by vivekreddynv** [comment]
Request for 202106

**Review by arlakshm — COMMENTED** [review]
as comments

File: tests/dump_tests/module_tests/copp_test.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+@patch("dump.plugins.copp.Copp.CONFIG_FILE", os.path.join(dump_test_input, "copp_cfg.json"))
+class TestCoppModule(unittest.TestCase):
+
+    def test_usr_cfg_trap_and_copp_cfg_file_grp(self):
+        '''
+        Scenario: A custom COPP_TRAP table entry is defined by the user and the relevant Trap Group is configured through the copp_cfg file
+        '''
+        params = {}
+        params[Copp.ARG_NAME] = "snmp"
+        params["namespace"] = ""
```
Comment (SuvarnaMeenakshi) [inline tests/dump_tests/module_tests/copp_test.py:51]: Can you add some test cases where namespace non-empty.

Comment (vivekreddynv) [reply tests/dump_tests/module_tests/copp_test.py:51]: Multi-asic tests are already covered in the match_infra tests [LINK](https://github.com/Azure/sonic-utilities/blob/master/tests/dump_tests/match_engine_test.py#L217) and also in the tests which verify the dump state tests which verify the "dump state cli" output. [LINK](https://github.com/Azure/sonic-utilities/blob/42664848239b4d41e28b86e86e79da532c292eef/tests/dump_tests/dump_state_test.py#L188) .

These UT's effectively capture the capability at both the match_infra and the top level cli levels. 

I don't think, adding non-default namespace tests here wouldn't add any value as it doesn't help verify anything specific to copp module implementation.

File: dump/plugins/copp.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
+import re
+from dump.match_infra import MatchEngine, MatchRequest
+from dump.helper import create_template_dict, handle_multiple_keys_matched_error
+from dump.helper import handle_error
+from .executor import Executor
+
+TRAP_ID_MAP = {
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py:7]: No, i've copied the table from copporch. [LINK](https://github.com/Azure/sonic-swss/blob/master/orchagent/copporch.cpp#L39). Seems copporch doesn't handle all the hostif traps defined in the SAI

File: dump/plugins/copp.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+            trap_id_key_cf, trap_group_cf = self.__find_trap_id_in_conf_file()
+        elif trap_id_key_db and not trap_group_db:
+            _, trap_group_cf = self.__find_trap_id_in_conf_file(trap_id_key_db.split("|")[-1], False)
+
+        # ------------------ Find any diff and fill the return dictionary with COPP_TRAP keys
+        if trap_id_key_db:
+            self.ret_temp["CONFIG_DB"]["keys"].append(trap_id_key_db)
+            self.copp_trap_cfg_key = trap_id_key_db.split("|")[-1]
+            id_in_file, _ = self.__find_trap_id_in_conf_file(self.copp_trap_cfg_key, False)
+            if id_in_file == trap_id_key_db:  # If any diff
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py:257]: id_in_file will either be trap_id_key_db or NULL. 

Eg: Let's say user provided trap_id as "ospfv6"

In Config DB:
```
 "COPP_TRAP":{
        "bgp": {
              "trap_ids": "ospfv6,ospf"
         }
  }
```

In Config File:
```
"COPP_TRAP": {
	    "bgp": {
		    "trap_ids": "bgp,bgpv6",
		    "trap_group": "queue4_group1"
	    }
}
```

So, here we found out trap_id_key_db is `bgp` given ospfv6 as the trap_id.  id_in_file is just a check in config file for any potential diff present in the COPP_TRAP|bgp key and thus id_in_file will either be "bgp" or NULL.

If it's NULL, we won't add it to the "CONFIG_FILE" section in the return template

File: dump/plugins/copp.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                self.ret_temp["CONFIG_FILE"]["keys"].append(trap_id_key_db)
+        elif trap_id_key_cf:
+            self.ret_temp["CONFIG_FILE"]["keys"].append(trap_id_key_cf)
+            self.copp_trap_cfg_key = trap_id_key_cf.split("|")[-1]
+            id_in_file, _ = self.__find_trap_id_in_db(self.copp_trap_cfg_key, False)
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py:262]: Sry, i'll rename this to id_in_db. 

And no, id_in_db can be trap_id_key_cf or NULL

Eg: Let's say user wanted a dump for trap_id: "ospfv6"

In Config File:
```
 "COPP_TRAP":{
        "bgp": {
              "trap_ids": "ospfv6,ospf"
         }
  }
```

In Config DB:
```
"COPP_TRAP": {
	    "bgp": {
		    "trap_ids": "bgp,bgpv6",
		    "trap_group": "queue4_group1"
	    }
}
```

In this case, after running `trap_id_key_db, trap_group_db = self.__find_trap_id_in_db()`, trap_id_key_db will still be empty as there is no direct way to find out the COPP_TRAP|bgp as the config DB entry doesn't have ospfv6,

We then find "trap_id_key_cf" from the file and the use that to find any diff present in the Config DB.

The complexity comes from the fact that there are two sources of truth for COPP config and you can provide a diff of any particular table (COPP_TRAP & COPP_GROUP) in any of the sources.

File: dump/plugins/copp.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+    def __fill_trap_group_in_conf_file(self, not_found_report=True):
+        req = MatchRequest(table=CFG_COPP_GROUP_TABLE_NAME, key_pattern=self.trap_group, ns=self.ns, file=Copp.CONFIG_FILE)
+        ret = self.match_engine.fetch(req)
+        key_tg = ""
+        if not ret["error"] and len(ret["keys"]) > 0:
+            key_tg = ret["keys"][0]
+            self.ret_temp["CONFIG_FILE"]["keys"].append(key_tg)
+            return True
+        elif not_found_report:
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py:295]: Yeah, but just to keep the implementation consistent with `__fill_trap_group_in_conf_db`, i've retained it.

File: dump/plugins/copp.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        req = MatchRequest(db="STATE_DB", table="COPP_GROUP_TABLE", key_pattern=self.trap_group)
+        ret = self.match_engine.fetch(req)
+        if not ret["error"] and len(ret["keys"]) > 0:
+            self.ret_temp["STATE_DB"]["keys"].append(ret["keys"][0])
+        else:
+            self.ret_temp["STATE_DB"]["tables_not_found"].append("COPP_GROUP_TABLE")
+
+    def handle_appl_db(self):
+        req = MatchRequest(db="APPL_DB", table=APP_COPP_TABLE_NAME, key_pattern="*", field="trap_ids",
+                           value=self.trap_id, match_entire_list=False, return_fields=["trap_group"])
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py:125]: I'm gettting a warning related to under-indentation when i change this.
Is this what you've suggested?
```
         req = MatchRequest(db="APPL_DB", table=APP_COPP_TABLE_NAME, key_pattern="*", field="trap_ids",
-                           value=self.trap_id, match_entire_list=False, return_fields=["trap_group"])
+              value=self.trap_id, match_entire_list=False, return_fields=["trap_group"])

```

File: dump/plugins/copp.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            if len(ret["keys"]) > 1:
+                err_str = "ERROR: Multiple COPP_TABLE Keys found for the trap_id {} in the APPL_DB, keys found: {}".format(self.trap_id, str(ret["keys"]))
+                handle_multiple_keys_matched_error(err_str, ret["keys"][0])
+            self.ret_temp["APPL_DB"]["keys"].append(ret["keys"][0])
+            tg = ret["return_values"][ret["keys"][0]]["trap_group"]
+        else:
+            self.ret_temp["APPL_DB"]["tables_not_found"].append(APP_COPP_TABLE_NAME)
+
+        if tg != self.trap_group and not self.trap_group and not tg:
+            err_str = "The Associated Trap_group for the trap_id found in APPL and CONFIG_DB/CONFIG_FILE did not match. In APPL_DB: {}, CONFIG_DB: {}".format(tg, self.trap_group)
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/copp.py]: Handled

**Review by dgsudharsan — APPROVED** [review]

**Comment by vivekreddynv** [comment]
@SuvarnaMeenakshi and @arlakshm Can you finish the review and sign-off?

**Comment by dgsudharsan** [comment]
@arlakshm  @SuvarnaMeenakshi  Can you please review and sign off?

**Review by arlakshm — APPROVED** [review]
lgtm

**Comment by dgsudharsan** [comment]
@SuvarnaMeenakshi  Can you please review and sign off?

**Comment by dgsudharsan** [comment]
@SuvarnaMeenakshi Can you please help here?

**Review by SuvarnaMeenakshi — APPROVED** [review]
